examples/ag_news.py

```diff
@@ -0,0 +1,114 @@
+
+
+def get_data() -> tuple:
+    """
+    Pulls the AG_NEWS dataset
+    and returns two tuples the first being the
+    training data and the second being the test
+    data. Each tuple contains the text and label
+    respsectively as numpy arrays.
+
+    """
+
+    train_iter, test_iter = AG_NEWS(split=("train", "test"))
+
+    train_text = []
+    train_labels = []
+    for label, text in train_iter:
+        train_labels.append(label)
+        train_text.append(text)
+
+    test_text = []
+    test_labels = []
+    for label, text in test_iter:
+        test_labels.append(label)
+        test_text.append(text)
+
+    train_text = np.array(train_text)
+    train_labels = np.array(train_labels)
+
+    test_text = np.array(test_text)
+    test_labels = np.array(test_labels)
+
+    train = (train_text, train_labels)
+    test = (test_text, test_labels)
+
+    return (train, test)
+
+
+def fit_model(
+    train_text: np.ndarray, train_labels: np.ndarray, distance_metric: str = "ncd"
+) -> KnnClassifier :
+    """
+    Fits a Knn-GZip compressor on the train
+    data and returns it.
+
+    Arguments:
+        train_text (np.ndarray): Training dataset as a numpy array.
+        train_labels (np.ndarray): Training labels as a numpy array.
+
+    Returns:
+        KnnClassifier: Trained Knn-Compressor model ready to make predictions.
+    """
+
+    compressor: BaseCompressor = GZipCompressor()
+    model: KnnClassifier = KnnClassifier(
+        compressor=compressor,
+        training_inputs=train_text,
+        training_labels=train_labels,
+        distance_metric=distance_metric,
+    )
+
+    return model
+
+def main() -> None:
+
+    print(f"Fetching data...")
+    ((train_text, train_labels), (test_text, test_labels)) = get_data()
+
+    print(f"Fitting model...")
+    model = fit_model(train_text, train_labels)
+    random_indicies = np.random.choice(test_text.shape[0], 1000, replace=False)  
+
+    sample_test_text = test_text[random_indicies]
+    sample_test_labels = test_labels[random_indicies]
+
+    print(f"Generating predictions...")
+    top_k = 1
+
+    # Here we use the `sampling_percentage` to save time
+    # at the expense of worse predictions. This 
+    # `sampling_percentage` selects a random % of training
+    # data to compare `sample_test_text` against rather
+    # than comparing it against the entire training dataset.
+    (distances, labels, similar_samples) = model.predict(
+        sample_test_text, top_k, sampling_percentage=0.01
+    )
+
+    print(classification_report(sample_test_labels, labels.reshape(-1)))
+
+
+
+
+if __name__ == "__main__":
+    main()
+
+
+#               precision    recall  f1-score   support
+
+#            1       0.24      0.09      0.13       258
+#            2       0.25      0.44      0.32       231
+#            3       0.26      0.06      0.09       244
+#            4       0.26      0.43      0.32       267
+
+#     accuracy                           0.26      1000
```

Comment:
The accuracy here is like a random guess - I think the way of drawing samples&labels need to be changed in order to have matched pairs of sample and label

Reply:
Yeah this is a bug in the np.random.choice() call when calculating the distance matrix. Adding "replace=False" should solve the random selection problem.

As for the accuracy being low, that's most likely because I'm only sampling 1% of the training data when making this classification report. I did that purely for speed in testing the codebase.

Reply:
Yeah I can re-run the test with "replace=False" but by skimming through the code I think the problem of mismatched text and labels still exists.
Sampling 1% of AGNews is using about 76 samples which is about 20-shot, whose accuracy should be higher than 5-shot at least. I think a simple sanity check is to see if the accuracy is higher than random guess - given anything > 3-shot is guaranteed to achieve accuracy greater than random guess.

Reply:
If there is somewhere the math is going wrong, it's probably in `npc_gzip/distance.py`, but this poor classification report should entirely be faulted on the 1% training data sampling. I would re-run on the entire dataset but it's going to take too long on my mac.

Is it worth removing the commented classification report altogether?

Reply:
> Yeah this is a bug in the np.random.choice() call when calculating the distance matrix. Adding "replace=False" should solve the random selection problem.
> 
> As for the accuracy being low, that's most likely because I'm only sampling 1% of the training data when making this classification report. I did that purely for speed in testing the codebase.

This has been fixed